crate_universe/private/crate.bzl

```
@@ -175,37 +175,42 @@ def _annotation(
         for patch in patches:
             _assert_absolute(patch)
 
+    def stringify_list(values):
+        if not values:
+            return values
+        return [str(x) for x in values]
+
     return json.encode((
         version,
         struct(
             additive_build_file = additive_build_file,
             additive_build_file_content = additive_build_file_content,
             alias_rule = parse_alias_rule(alias_rule),
-            build_script_data = build_script_data,
-            build_script_tools = build_script_tools,
+            build_script_data = stringify_list(build_script_data),
+            build_script_tools = stringify_list(build_script_tools),
             build_script_data_glob = build_script_data_glob,
-            build_script_deps = build_script_deps,
+            build_script_deps = stringify_list(build_script_deps),
             build_script_env = build_script_env,
-            build_script_proc_macro_deps = build_script_proc_macro_deps,
+            build_script_proc_macro_deps = stringify_list(build_script_proc_macro_deps),
             build_script_rundir = build_script_rundir,
             build_script_rustc_env = build_script_rustc_env,
-            build_script_toolchains = build_script_toolchains,
-            compile_data = compile_data,
+            build_script_toolchains = stringify_list(build_script_toolchains),
+            compile_data = stringify_list(compile_data),
             compile_data_glob = compile_data_glob,
             crate_features = crate_features,
-            data = data,
+            data = stringify_list(data),
             data_glob = data_glob,
-            deps = deps,
+            deps = stringify_list(deps),
             extra_aliased_targets = extra_aliased_targets,
             gen_binaries = gen_binaries,
             disable_pipelining = disable_pipelining,
             gen_build_script = gen_build_script,
             patch_args = patch_args,
             patch_tool = patch_tool,
             patches = patches,
```

```suggestion
            patches = stringify_list(patches),
```

This is also needed here.